### PR TITLE
[caffe] Add new port

### PR DESCRIPTION
**Pull request by vlj**
This port is adding caffe-1.0 to vcpkg.
While caffe2 is newer it's not completly backward compatible which motivates this port. Additionnaly Caffe has an opencl branch (this is so far the only Deep Learning Framework with an opencl backend so far) which I'd like to bring to vcpkg through features.

Initially there is no CUDA/OpenCV/Python/Matlab support.

### CODE REVIEW

**Comment by vlj** [comment]
Added python support

**Comment by jasjuang** [comment]
Just wondering did you check out @willyd 's port https://github.com/BVLC/caffe/issues/5674? His port works with OpenCV and CUDA but he never submitted a PR to vcpkg (not sure why). Can you look into it and update this PR?

**Comment by willyd** [comment]
@vjl This port has been on my TODO list for a while now. I have one that works but is different from yours in some ways:

- Mine is based off the caffe/windows branch
- Mine supports CUDA and OpenCV
- Your supports Python. I wonder how this is possible, vcpkg overwrites the PATH variable so CMake can't find my Anaconda Python. 
- Mine supports Dynamic linking (Also supports static linking in theory, but I haven't tried to build it yet)

I haven't submitted a PR since I needed to make some changes to the caffe code base for the port to work and I still need to find some time to merge the changes in the caffe windows branch.

Would you like to collaborate on this?

**Comment by vlj** [comment]
I added CUDA and OpenCV using @willyd branch. I checked both and OpenCV works, but Cuda fails with the following :
"fatal error C1189: #error:  -- unsupported Microsoft Visual Studio version! Only the versions 2012, 2013, 2015 and 2017 are supported!"

I have vanilla Python 3 installed and I didn't get issue at all. I suspect vcpkg uses it's own python, but relies on Python_packages_path envvar to find Numpy.

For me dynamic linking builds but doesn't work, caffe executable fails because protobuf is misleaded by a static symbol found twice. This is an instance of this bug : https://github.com/BVLC/caffe/issues/1917

The branch is based on the 1.0 release with patch backported from the Windows branch.

**Comment by jasjuang** [comment]
@vlj I suspect you are using VS2017 15.5, unfortunately you need to downgrade it to VS2017 15.4 for CUDA to work. See https://devtalk.nvidia.com/default/topic/1027299/cuda-setup-and-installation/cuda-9-failed-to-support-the-latest-visual-studio-2017-version-15-5/

**Comment by willyd** [comment]
@vlj There is a detail writeup on a to workaround the CUDA issue here: https://github.com/BVLC/caffe/issues/5654#issuecomment-354539618

As @jasjuang mentionned the safest route is probably to downgrade until NVIDIA releases an update.

**Comment by willyd** [comment]
> I have vanilla Python 3 installed and I didn't get issue at all. I suspect vcpkg uses it's own python, but relies on Python_packages_path envvar to find Numpy.

I guess CMake looks up the Python registry entries. This won't work with conda environments and virtual envs. Sad.

> For me dynamic linking builds but doesn't work, caffe executable fails because protobuf is misleaded by a static symbol found twice. This is an instance of this bug : BVLC/caffe#1917

Yeah, I had to fiddle a lot with the build to get this to work. Changed caffeproto from STATIC to OBJECT and changed the protoc command line call in Protobuf.cmake. See this commit https://github.com/willyd/caffe/commit/8aee77a3b8983310dab023f94d427c18caebfa48

**Comment by vlj** [comment]
I updated the port so that it uses dynamic instead of static, using @willyd tips (declaring proto as OBJECT instead of STATIC). I also had to use the same tricks for pycaffe.

With these modifications I got Deepdream script working : https://www.alanzucconi.com/2016/05/25/generating-deep-dreams/
natively on Windows with dynamic linked caffe.

Unfortunatly the static version doesn't work, it looks like it strips static symbol when linking proto (hence the layers descriptions are not found when pycaffe is loaded and it's not possible to use any caffeproto file). I will import the export related code in static build in the Windows branch of caffe I guess.

**Comment by willyd** [comment]
@vjl This is great! Use the /wholearchive linker flag. This way you won't need to use the pragma trick that I implemented in the windows branch. I am going to drop this along VS 2013 support anyway.

**Comment by vlj** [comment]
Actually I had to revert back to using static linking to enable CUDA support ; OBJECT doesn't accept *.cu file.
I got Deepdream python script working however it gives me some strange result with the sky picture, as if it was random :

**Comment by willyd** [comment]
@vjl It was only necessary to use `OBJECT` for the caffeproto library. This library does not include any .cu files. You have to use the `WINDOWS_EXPORT_ALL_SYMBOLS` CMake trick though. 

Can't comment on the DeepDream result I always wanted to try it but never got to it. I usually try MNIST or just a simple linear regression to see if everything is working fine.

**Comment by vlj** [comment]
I opened an issue here : https://github.com/BVLC/caffe/issues/6192

**Comment by ras0219-msft** [comment]
LMK when this is ready to be merged :)

**Comment by vlj** [comment]
I'm a bit puzzled since the caffe prebuilt binaries from this link : https://github.com/BVLC/caffe/tree/windows
works properly with the deepdream script. 

Unit tests mostly pass.
There's some patch I didn't import like the one moving Template from headers to cpp files (but I think using /WHOLEARCHIVE flag is equivalent) and some fix to MakeTempDirectory but I'm not sure it's used for anything but tests. There's also leveldb which is missing but it doesn't seem to be used to load caffemodel files.
I'm investigating

**Comment by vlj** [comment]
It looks like protobuf fails to load the caffemodel file, I don't know why. I modified the proto definition by adding an underscore to STRICT and PERMISSIVE, but removing them and manually changing the var definition in the autogenerated file doesn't seem to have any influence.

**Comment by vlj** [comment]
I fixed the issue, actually I missed a hunk (open needs the binary flag to read caffeproto files).
The branch should be ready to merge I think.

**Comment by vlj** [comment]
Feedback ?

**Comment by jasjuang** [comment]
Looks great to me! can't wait to see it to be merged

File: ports/caffe/0001-Use-vcpkg-dependencies.patch [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+ 
+ # ---[ Google-protobuf
+ include(cmake/ProtoBuf.cmake)
+ 
+ # ---[ HDF5
+-find_package(HDF5 COMPONENTS HL REQUIRED)
+-list(APPEND Caffe_INCLUDE_DIRS PUBLIC ${HDF5_INCLUDE_DIRS})
+-list(APPEND Caffe_LINKER_LIBS PUBLIC ${HDF5_LIBRARIES} ${HDF5_HL_LIBRARIES})
++find_package(hdf5 COMPONENTS HL REQUIRED)
++list(APPEND Caffe_LINKER_LIBS PUBLIC hdf5-shared hdf5_hl-shared)
```
Comment (willyd) [inline ports/caffe/0001-Use-vcpkg-dependencies.patch]: x64-windows-static would require to link to hdf5-static here.

File: ports/caffe/0002-Fix-permissive-strict.patch [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+--- a/src/caffe/proto/caffe.proto
++++ b/src/caffe/proto/caffe.proto
+@@ -293,9 +293,9 @@ message ParamSpec {
+   optional DimCheckMode share_mode = 2;
+   enum DimCheckMode {
+     // STRICT (default) requires that num, channels, height, width each match.
+-    STRICT = 0;
++    STRICT_ = 0;
+     // PERMISSIVE requires only the count (num*channels*height*width) to match.
+-    PERMISSIVE = 1;
```
Comment (willyd) [inline ports/caffe/0002-Fix-permissive-strict.patch]: We should use the NO_STRICT compile definition instead.

File: ports/caffe/0002-Fix-permissive-strict.patch [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
++    PERMISSIVE_ = 1;
+   }
+ 
+   // The multiplier on the global learning rate for this parameter.
+@@ -1268,8 +1268,8 @@ message V1LayerParameter {
+   repeated string param = 1001;
+   repeated DimCheckMode blob_share_mode = 1002;
+   enum DimCheckMode {
+-    STRICT = 0;
+-    PERMISSIVE = 1;
```
Comment (willyd) [inline ports/caffe/0002-Fix-permissive-strict.patch]: Ditto.

File: ports/caffe/0004-Export-fix.patch [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+diff --git a/python/CMakeLists.txt b/python/CMakeLists.txt
+index c53299d2..20af0dcf 100644
+--- a/python/CMakeLists.txt
++++ b/python/CMakeLists.txt
+@@ -10,6 +10,9 @@ caffe_default_properties(pycaffe)
+ set_target_properties(pycaffe PROPERTIES PREFIX "" OUTPUT_NAME "_caffe")
+ target_include_directories(pycaffe PUBLIC ${PYTHON_INCLUDE_DIRS} ${NUMPY_INCLUDE_DIR})
+ target_link_libraries(pycaffe PUBLIC ${Caffe_LINK} ${PYTHON_LIBRARIES})
++set_target_properties(pycaffe PROPERTIES LINK_FLAGS "/WHOLEARCHIVE:caffe.lib")
```
Comment (willyd) [inline ports/caffe/0004-Export-fix.patch]: Would a generator expression work here: `/WHOLEARCHIVE:$<TARGET_FILE_NAME:caffe>`?

File: ports/caffe/0004-Export-fix.patch [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+diff --git a/src/gtest/CMakeLists.txt b/src/gtest/CMakeLists.txt
+index e98254af..827e0176 100644
+--- a/src/gtest/CMakeLists.txt
++++ b/src/gtest/CMakeLists.txt
+@@ -2,7 +2,9 @@ add_library(gtest STATIC EXCLUDE_FROM_ALL gtest.h gtest-all.cpp)
+ caffe_default_properties(gtest)
+ target_include_directories(gtest PUBLIC ${Caffe_SRC_DIR})
+ target_compile_definitions(gtest PUBLIC -DGTEST_USE_OWN_TR1_TUPLE)
+-
++set_target_properties(gtest PROPERTIES LINK_FLAGS_RELEASE "/WHOLEARCHIVE:caffe")
```
Comment (willyd) [inline ports/caffe/0004-Export-fix.patch]: Should probably be in the LINK_INTERFACE_LIBRARIES of the caffe lib instead.

File: ports/caffe/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        -DBUILD_python=OFF
+    )
+
+vcpkg_install_cmake()
+
+# Remove folders from install
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/share)
+
+#file(COPY ${CURRENT_PACKAGES_DIR}/debug/lib/caffe-d.dll DESTINATION ${CURRENT_PACKAGES_DIR}/debug/bin)
```
Comment (willyd) [inline ports/caffe/portfile.cmake]: You should use vcpkg_fixup_cmake_targets too!

**Comment by willyd** [comment]
@vlj Please see my comments on your patches.

File: ports/caffe/portfile.cmake [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    PATCHES
+    ${CMAKE_CURRENT_LIST_DIR}/0001-Use-vcpkg-dependencies.patch
+    ${CMAKE_CURRENT_LIST_DIR}/0002-Fix-permissive-strict.patch
+    ${CMAKE_CURRENT_LIST_DIR}/0003-MSVC-fixes.patch
+    ${CMAKE_CURRENT_LIST_DIR}/0004-Export-fix.patch
+    ${CMAKE_CURRENT_LIST_DIR}/0005-Disable-examples.patch
+    ${CMAKE_CURRENT_LIST_DIR}/0006-Do-not-rename-caffe.bin.exe-to-avoid-confusing-ninja.patch
+    ${CMAKE_CURRENT_LIST_DIR}/0007-Fix-dependencies.patch
+    ${CMAKE_CURRENT_LIST_DIR}/0008-Make-proto-OBJECT-to-avoid-protobuf-loading-caffe-pr.patch
+    ${CMAKE_CURRENT_LIST_DIR}/0009-Fix-Cuda.patch
```
Comment (NancyLi1013) [inline ports/caffe/portfile.cmake]: Can these patches be submitted upstream?

Besides this, `vcpkg_apply_patches()` is deprecated, if there are some patches needed by this port, please use `vcpkg_from_github()` instead.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by dan-shaw** [comment]
I’m going to close this PR because it’s been a long time since any update has been made. If this is something you’d like to see merged and work is being done, please reopen.

**Pull request closed without merging**